File: team417/src/main/java/org/firstinspires/ftc/team417/SparkFunOTOSCorrected.java
package org.firstinspires.ftc.team417;

import com.qualcomm.hardware.sparkfun.SparkFunOTOS;
import com.qualcomm.robotcore.hardware.I2cDeviceSynch;
import com.qualcomm.robotcore.hardware.configuration.annotations.DeviceProperties;
import com.qualcomm.robotcore.hardware.configuration.annotations.I2cDeviceType;
import java.util.*;

@I2cDeviceType
@DeviceProperties(
        name = "SparkFun OTOS Corrected",
        xmlTag = "SparkFunOTOS2",
        description = "SparkFun Qwiic Optical Tracking Odometry Sensor Corrected"
)
public class SparkFunOTOSCorrected extends SparkFunOTOS {
    public SparkFunOTOSCorrected(I2cDeviceSynch deviceClient) {
        super(deviceClient);
    }

    // Modified version of poseToRegs to fix pose setting issue
    // see https://discord.com/channels/225450307654647808/1246977443030368349/1271702497659977760
    @Override
    protected void poseToRegs(byte[] rawData, Pose2D pose, double xyToRaw, double hToRaw) {
        // Convert pose units to raw data
        short rawX = (short) (_distanceUnit.toMeters(pose.x) * xyToRaw);
        short rawY = (short) (_distanceUnit.toMeters(pose.y) * xyToRaw);
        short rawH = (short) (_angularUnit.toRadians(pose.h) * hToRaw);

        // Store raw data in buffer
        rawData[0] = (byte) (rawX & 0xFF);
        rawData[1] = (byte) ((rawX >> 8) & 0xFF);
        rawData[2] = (byte) (rawY & 0xFF);
        rawData[3] = (byte) ((rawY >> 8) & 0xFF);
        rawData[4] = (byte) (rawH & 0xFF);
        rawData[5] = (byte) ((rawH >> 8) & 0xFF);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/pedroPathing/localization/SparkFunOTOSCorrected.kt
package com.acmerobotics.roadrunner.ftc

import com.qualcomm.hardware.sparkfun.SparkFunOTOS
import com.qualcomm.robotcore.hardware.I2cDeviceSynch
import com.qualcomm.robotcore.hardware.configuration.annotations.DeviceProperties
import com.qualcomm.robotcore.hardware.configuration.annotations.I2cDeviceType
import java.util.*
@I2cDeviceType
@DeviceProperties(
    name = "SparkFun OTOS Corrected",
    xmlTag = "SparkFunOTOS2",
    description = "SparkFun Qwiic Optical Tracking Odometry Sensor Corrected"
)
class SparkFunOTOSCorrected(deviceClient: I2cDeviceSynch) : SparkFunOTOS(deviceClient) {
    /**
     * Gets only the position and velocity measured by the
     * OTOS in a single burst read
     * @param pos Position measured by the OTOS
     * @param vel Velocity measured by the OTOS
     */
    fun getPosVel(pos: Pose2D, vel: Pose2D) {
        // Read all pose registers
        val rawData = deviceClient.read(REG_POS_XL.toInt(), 12)

        // Convert raw data to pose units
        pos.set(regsToPose(Arrays.copyOfRange(rawData, 0, 6), INT16_TO_METER, INT16_TO_RAD))
        vel.set(regsToPose(Arrays.copyOfRange(rawData, 6, 12), INT16_TO_MPS, INT16_TO_RPS))
    }

    // Modified version of poseToRegs to fix pose setting issue
    // see https://discord.com/channels/225450307654647808/1246977443030368349/1271702497659977760
    override fun poseToRegs(rawData: ByteArray, pose: Pose2D, xyToRaw: Double, hToRaw: Double) {
        // Convert pose units to raw data
        val rawX = (_distanceUnit.toMeters(pose.x) * xyToRaw).toInt().toShort()
        val rawY = (_distanceUnit.toMeters(pose.y) * xyToRaw).toInt().toShort()
        val rawH = (_angularUnit.toRadians(pose.h) * hToRaw).toInt().toShort()

        // Store raw data in buffer
        rawData[0] = (rawX.toInt() and 0xFF).toByte()
        rawData[1] = ((rawX.toInt() shr 8) and 0xFF).toByte()
        rawData[2] = (rawY.toInt() and 0xFF).toByte()
        rawData[3] = ((rawY.toInt() shr 8) and 0xFF).toByte()
        rawData[4] = (rawH.toInt() and 0xFF).toByte()
        rawData[5] = ((rawH.toInt() shr 8) and 0xFF).toByte()
    }
}